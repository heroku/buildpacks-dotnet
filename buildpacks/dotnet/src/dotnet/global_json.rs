use semver::VersionReq;
use serde::Deserialize;
use std::convert::TryFrom;
use std::str::FromStr;

/// Represents the root structure of a global.json file.
#[derive(Deserialize)]
pub(crate) struct GlobalJson {
    pub(crate) sdk: Option<SdkConfig>,
}

/// Represents the SDK configuration in a global.json file.
#[derive(Deserialize)]
pub(crate) struct SdkConfig {
    version: String,
    #[serde(rename = "rollForward")]
    roll_forward: Option<String>,
}

impl FromStr for GlobalJson {
    type Err = serde_json::Error;

    fn from_str(contents: &str) -> Result<Self, Self::Err> {
        serde_json::from_str::<GlobalJson>(contents)
    }
}

impl TryFrom<SdkConfig> for VersionReq {
    type Error = semver::Error;

    // TODO: Factor in pre-release logic
    fn try_from(sdk_config: SdkConfig) -> Result<Self, Self::Error> {
        let version = &sdk_config.version;
        let roll_forward = sdk_config.roll_forward.as_deref();

        let version_req_str = match roll_forward {
            Some("patch" | "latestPatch") => format!("~{version}"),
            Some("feature" | "latestFeature") => format!(
                "~{}",
                version.split('.').take(2).collect::<Vec<_>>().join(".")
            ),
            Some("minor" | "latestMinor") => format!(
                "^{}",
                version.split('.').take(2).collect::<Vec<_>>().join(".")
            ),
            Some("major" | "latestMajor") => "*".to_string(),
            Some("disable") => format!("={version}"),
            _ => version.clone(),
        };
        VersionReq::parse(&version_req_str)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use semver::VersionReq;
    use std::str::FromStr;

    #[test]
    fn test_construct_version_req_from_sdk() {
        #[derive(Debug)]
        struct TestCase {
            version: &'static str,
            roll_forward: Option<&'static str>,
            expected: &'static str,
        }

        let test_cases = [
            TestCase {
                version: "6.0.100",
                roll_forward: Some("patch"),
                expected: "~6.0.100",
            },
            TestCase {
                version: "6.0.100",
                roll_forward: Some("latestPatch"),
                expected: "~6.0.100",
            },
            TestCase {
                version: "6.0.100",
                roll_forward: Some("feature"),
                expected: "~6.0",
            },
            TestCase {
                version: "6.0.100",
                roll_forward: Some("latestFeature"),
                expected: "~6.0",
            },
            TestCase {
                version: "6.0.100",
                roll_forward: Some("minor"),
                expected: "^6.0",
            },
            TestCase {
                version: "6.0.100",
                roll_forward: Some("latestMinor"),
                expected: "^6.0",
            },
            TestCase {
                version: "6.0.100",
                roll_forward: Some("major"),
                expected: "*",
            },
            TestCase {
                version: "6.0.100",
                roll_forward: Some("latestMajor"),
                expected: "*",
            },
            TestCase {
                version: "6.0.100",
                roll_forward: Some("disable"),
                expected: "=6.0.100",
            },
            TestCase {
                version: "6.0.100-rc.1.12345.1",
                roll_forward: Some("disable"),
                expected: "=6.0.100-rc.1.12345.1",
            },
            TestCase {
                version: "6.0.100",
                roll_forward: None,
                expected: "^6.0.100",
            },
            TestCase {
                version: "6.0.100-rc.1.12345.1",
                roll_forward: None,
                expected: "^6.0.100-rc.1.12345.1",
            },
        ];

        for case in &test_cases {
            let sdk_config = SdkConfig {
                version: case.version.to_string(),
                roll_forward: case.roll_forward.map(ToString::to_string),
            };
            let result = VersionReq::try_from(sdk_config).unwrap();
            assert_eq!(
                result.to_string(),
                case.expected,
                "Failed for case: {case:?}"
            );
        }
    }

    #[test]
    fn test_parse_global_json_with_sdk() {
        let json_content = r#"
        {
            "sdk": {
                "version": "6.0.100",
                "rollForward": "latestMinor"
            }
        }
        "#;

        let global_json = GlobalJson::from_str(json_content).unwrap();
        let version_req = VersionReq::try_from(global_json.sdk.unwrap()).unwrap();
        assert_eq!(version_req, VersionReq::parse("^6.0").unwrap());
    }

    #[test]
    fn test_parse_global_json_without_sdk_rollforward() {
        let sdk_config = SdkConfig {
            version: "6.0.100".to_string(),
            roll_forward: None,
        };
        let version_req = VersionReq::try_from(sdk_config).unwrap();
        assert_eq!(version_req, VersionReq::parse("6.0.100").unwrap());
    }

    #[test]
    fn test_parse_empty_global_json() {
        let json_content = r"
        {
        }
        ";
        let global_json = GlobalJson::from_str(json_content).unwrap();
        assert!(global_json.sdk.is_none());
    }
}
