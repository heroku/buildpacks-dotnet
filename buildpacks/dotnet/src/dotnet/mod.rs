pub(crate) mod global_json;
pub(crate) mod project;
pub(crate) mod runtime_identifier;
pub(crate) mod solution;
pub(crate) mod target_framework_moniker;
