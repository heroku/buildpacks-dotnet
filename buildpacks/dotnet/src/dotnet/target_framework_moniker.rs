use semver::VersionReq;
use std::convert::TryFrom;
use std::str::FromStr;

#[derive(Debug)]
pub(crate) enum ParseTargetFrameworkError {
    InvalidFormat(String),
    UnsupportedOSTfm(String),
}

#[derive(Debug)]
pub(crate) struct TargetFrameworkMoniker {
    pub(crate) version_part: String,
}

impl FromStr for TargetFrameworkMoniker {
    type Err = ParseTargetFrameworkError;

    fn from_str(tfm: &str) -> Result<Self, Self::Err> {
        let valid_prefixes = ["net"];

        if tfm.len() < 4 {
            return Err(ParseTargetFrameworkError::InvalidFormat(tfm.to_string()));
        }

        let prefix = &tfm[..3];
        let rest = &tfm[3..];

        if !valid_prefixes.contains(&prefix) || rest.is_empty() {
            return Err(ParseTargetFrameworkError::InvalidFormat(tfm.to_string()));
        }

        let parts: Vec<&str> = rest.split('-').collect();
        if parts.len() > 1 {
            return Err(ParseTargetFrameworkError::UnsupportedOSTfm(tfm.to_string()));
        }

        let version_part = parts[0]
            .split('.')
            .filter(|part| part.chars().all(char::is_numeric))
            .collect::<Vec<&str>>()
            .join(".");

        if version_part.is_empty() || !rest.chars().all(|c| c.is_numeric() || c == '.') {
            return Err(ParseTargetFrameworkError::InvalidFormat(tfm.to_string()));
        }

        Ok(TargetFrameworkMoniker { version_part })
    }
}

impl TryFrom<&TargetFrameworkMoniker> for VersionReq {
    type Error = semver::Error;

    fn try_from(tf: &TargetFrameworkMoniker) -> Result<Self, Self::Error> {
        VersionReq::from_str(&format!("^{}", tf.version_part))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use semver::VersionReq;
    use std::convert::TryFrom;

    #[test]
    fn test_parse_net6_0() {
        let tfm = "net6.0";
        let target_framework = tfm.parse::<TargetFrameworkMoniker>().unwrap();
        let expected = VersionReq::from_str("^6.0").unwrap();
        assert_eq!(VersionReq::try_from(&target_framework).unwrap(), expected);
    }

    #[test]
    fn test_parse_net7_0() {
        let tfm = "net7.0";
        let target_framework = tfm.parse::<TargetFrameworkMoniker>().unwrap();
        let expected = VersionReq::from_str("^7.0").unwrap();
        assert_eq!(VersionReq::try_from(&target_framework).unwrap(), expected);
    }

    #[test]
    fn test_parse_net8_0() {
        let tfm = "net8.0";
        let target_framework = tfm.parse::<TargetFrameworkMoniker>().unwrap();
        let expected = VersionReq::from_str("^8.0").unwrap();
        assert_eq!(VersionReq::try_from(&target_framework).unwrap(), expected);
    }

    #[test]
    fn test_parse_invalid_empty() {
        let tfm = "";
        assert!(matches!(
            tfm.parse::<TargetFrameworkMoniker>(),
            Err(ParseTargetFrameworkError::InvalidFormat(_))
        ));
    }

    #[test]
    fn test_parse_invalid_non_numeric() {
        let tfm = String::from("netcoreapp");
        assert!(matches!(
            tfm.parse::<TargetFrameworkMoniker>(),
            Err(ParseTargetFrameworkError::InvalidFormat(_))
        ));
    }

    #[test]
    fn test_parse_invalid_malformed_version() {
        let tfm = "net6.x";
        assert!(matches!(
            tfm.parse::<TargetFrameworkMoniker>(),
            Err(ParseTargetFrameworkError::InvalidFormat(_))
        ));
    }

    #[test]
    fn test_parse_unsupported_os() {
        let tfm = "net6.0-ios15.0";
        assert!(matches!(
            tfm.parse::<TargetFrameworkMoniker>(),
            Err(ParseTargetFrameworkError::UnsupportedOSTfm(_))
        ));
    }
}
