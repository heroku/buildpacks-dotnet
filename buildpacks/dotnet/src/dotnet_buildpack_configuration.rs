use std::fmt;
use std::str::FromStr;

#[derive(Debug, PartialEq)]
pub(crate) struct DotnetBuildpackConfiguration {
    pub(crate) build_configuration: Option<String>,
    pub(crate) execution_environment: ExecutionEnvironment,
    pub(crate) msbuild_verbosity_level: Option<VerbosityLevel>,
}

#[derive(Debug, PartialEq)]
pub(crate) enum DotnetBuildpackConfigurationError {
    ExecutionEnvironmentError(ExecutionEnvironmentError),
    InvalidMsbuildVerbosityLevel(String),
}

impl TryFrom<&libcnb::Env> for DotnetBuildpackConfiguration {
    type Error = DotnetBuildpackConfigurationError;

    fn try_from(env: &libcnb::Env) -> Result<Self, Self::Error> {
        Ok(Self {
            build_configuration: env.get_string_lossy("BUILD_CONFIGURATION"),
            execution_environment: env
                .get_string_lossy("CNB_EXEC_ENV")
                .as_deref()
                .map_or_else(
                    || Ok(ExecutionEnvironment::Production),
                    ExecutionEnvironment::from_str,
                )
                .map_err(DotnetBuildpackConfigurationError::ExecutionEnvironmentError)?,
            msbuild_verbosity_level: detect_msbuild_verbosity_level(env).transpose()?,
        })
    }
}

fn detect_msbuild_verbosity_level(
    env: &libcnb::Env,
) -> Option<Result<VerbosityLevel, DotnetBuildpackConfigurationError>> {
    env.get("MSBUILD_VERBOSITY_LEVEL")
        .map(|value| value.to_string_lossy())
        .map(|value| match value.to_lowercase().as_str() {
            "q" | "quiet" => Ok(VerbosityLevel::Quiet),
            "m" | "minimal" => Ok(VerbosityLevel::Minimal),
            "n" | "normal" => Ok(VerbosityLevel::Normal),
            "d" | "detailed" => Ok(VerbosityLevel::Detailed),
            "diag" | "diagnostic" => Ok(VerbosityLevel::Diagnostic),
            _ => Err(
                DotnetBuildpackConfigurationError::InvalidMsbuildVerbosityLevel(value.to_string()),
            ),
        })
}

#[derive(Debug, PartialEq)]
pub(crate) enum ExecutionEnvironment {
    Production,
    Test,
}

impl FromStr for ExecutionEnvironment {
    type Err = ExecutionEnvironmentError;

    fn from_str(cnb_exec_env: &str) -> Result<Self, Self::Err> {
        match cnb_exec_env {
            "production" => Ok(ExecutionEnvironment::Production),
            "test" => Ok(ExecutionEnvironment::Test),
            _ => Err(ExecutionEnvironmentError::UnsupportedExecutionEnvironment(
                cnb_exec_env.to_string(),
            )),
        }
    }
}

#[derive(Debug, PartialEq)]
pub(crate) enum ExecutionEnvironmentError {
    UnsupportedExecutionEnvironment(String),
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub(crate) enum VerbosityLevel {
    Quiet,
    Minimal,
    Normal,
    Detailed,
    Diagnostic,
}

impl fmt::Display for VerbosityLevel {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            VerbosityLevel::Quiet => write!(f, "quiet"),
            VerbosityLevel::Minimal => write!(f, "minimal"),
            VerbosityLevel::Normal => write!(f, "normal"),
            VerbosityLevel::Detailed => write!(f, "detailed"),
            VerbosityLevel::Diagnostic => write!(f, "diagnostic"),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use libcnb::Env;

    fn create_env(variables: &[(&str, &str)]) -> Env {
        let mut env = Env::new();
        for &(key, value) in variables {
            env.insert(key, value);
        }
        env
    }

    #[test]
    fn test_default_buildpack_configuration() {
        let env = create_env(&[]);
        let result = DotnetBuildpackConfiguration::try_from(&env).unwrap();

        assert_eq!(
            result,
            DotnetBuildpackConfiguration {
                build_configuration: None,
                execution_environment: ExecutionEnvironment::Production,
                msbuild_verbosity_level: None
            }
        );
    }

    #[test]
    fn test_buildpack_configuration_test_execution_environment() {
        let env = create_env(&[("CNB_EXEC_ENV", "test")]);
        let result = DotnetBuildpackConfiguration::try_from(&env).unwrap();

        assert_eq!(result.execution_environment, ExecutionEnvironment::Test);
    }

    #[test]
    fn test_parse_execution_environment() {
        let cases = [
            ("production", Ok(ExecutionEnvironment::Production)),
            ("test", Ok(ExecutionEnvironment::Test)),
            (
                "foo",
                Err(ExecutionEnvironmentError::UnsupportedExecutionEnvironment(
                    "foo".to_string(),
                )),
            ),
        ];
        for (input, expected) in cases {
            assert_eq!(ExecutionEnvironment::from_str(input), expected);
        }
    }

    #[test]
    fn test_detect_msbuild_verbosity_level() {
        let cases = [
            ("q", Ok(VerbosityLevel::Quiet)),
            ("quiet", Ok(VerbosityLevel::Quiet)),
            ("minimal", Ok(VerbosityLevel::Minimal)),
            ("m", Ok(VerbosityLevel::Minimal)),
            ("normal", Ok(VerbosityLevel::Normal)),
            ("n", Ok(VerbosityLevel::Normal)),
            ("detailed", Ok(VerbosityLevel::Detailed)),
            ("d", Ok(VerbosityLevel::Detailed)),
            ("diagnostic", Ok(VerbosityLevel::Diagnostic)),
            ("diag", Ok(VerbosityLevel::Diagnostic)),
            (
                "invalid",
                Err(
                    DotnetBuildpackConfigurationError::InvalidMsbuildVerbosityLevel(
                        "invalid".to_string(),
                    ),
                ),
            ),
        ];

        for (input, expected) in cases {
            let env = create_env(&[("MSBUILD_VERBOSITY_LEVEL", input)]);
            let result = detect_msbuild_verbosity_level(&env);
            assert_eq!(result, Some(expected));
        }
        assert!(detect_msbuild_verbosity_level(&Env::new()).is_none());
    }
}
