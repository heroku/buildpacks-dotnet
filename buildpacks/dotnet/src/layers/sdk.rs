use crate::{DotnetBuildpack, DotnetBuildpackError};
use bullet_stream::global::print;
use bullet_stream::style;
use inventory::artifact::Artifact;
use inventory::checksum::Checksum;
use libcnb::data::layer_name;
use libcnb::layer::{
    CachedLayerDefinition, EmptyLayerCause, InvalidMetadataAction, LayerRef, LayerState,
    RestoredLayerAction,
};
use libherokubuildpack::download::{download_file, DownloadError};
use libherokubuildpack::inventory;
use libherokubuildpack::tar::decompress_tarball;
use semver::Version;
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha512};
use std::env::temp_dir;
use std::fs::{self, File};
use std::path::Path;
use std::thread;
use std::time::Duration;

#[derive(Serialize, Deserialize)]
pub(crate) struct SdkLayerMetadata {
    artifact: Artifact<Version, Sha512, Option<()>>,
}

pub(crate) enum CustomCause {
    Ok,
    DifferentSdkArtifact(Artifact<Version, Sha512, Option<()>>),
}

const MAX_RETRIES: u8 = 4;

pub(crate) fn handle(
    context: &libcnb::build::BuildContext<DotnetBuildpack>,
    available_at_launch: bool,
    artifact: &Artifact<Version, Sha512, Option<()>>,
) -> Result<LayerRef<DotnetBuildpack, (), CustomCause>, libcnb::Error<DotnetBuildpackError>> {
    let sdk_layer = context.cached_layer(
        layer_name!("sdk"),
        CachedLayerDefinition {
            build: true,
            launch: available_at_launch,
            invalid_metadata_action: &|_| InvalidMetadataAction::DeleteLayer,
            restored_layer_action: &|metadata: &SdkLayerMetadata, _path| {
                if metadata.artifact == *artifact {
                    (RestoredLayerAction::KeepLayer, CustomCause::Ok)
                } else {
                    (
                        RestoredLayerAction::DeleteLayer,
                        CustomCause::DifferentSdkArtifact(metadata.artifact.clone()),
                    )
                }
            },
        },
    )?;

    print::bullet("SDK installation");

    match sdk_layer.state {
        LayerState::Restored { .. } => {
            print::sub_bullet(format!("Reusing cached SDK (version {})", artifact.version));
        }
        LayerState::Empty { ref cause } => {
            if let EmptyLayerCause::RestoredLayerAction {
                cause: CustomCause::DifferentSdkArtifact(old_artifact),
            } = cause
            {
                print::sub_bullet(format!(
                    "Deleting cached .NET SDK (version {})",
                    old_artifact.version
                ));
            }

            sdk_layer.write_metadata(SdkLayerMetadata {
                artifact: artifact.clone(),
            })?;

            let mut log_background_bullet = print::sub_start_timer(format!(
                "Downloading SDK from {}",
                style::url(artifact.clone().url)
            ));

            let tarball_path = temp_dir().join("dotnetsdk.tar.gz");
            let mut download_attempts = 0;
            while download_attempts <= MAX_RETRIES {
                match download_file(&artifact.url, &tarball_path) {
                    Err(DownloadError::IoError(error)) if download_attempts < MAX_RETRIES => {
                        let sub_bullet = log_background_bullet.cancel(format!("{error}"));
                        download_attempts += 1;
                        thread::sleep(Duration::from_secs(1));
                        log_background_bullet = sub_bullet.start_timer("Retrying");
                    }
                    result => {
                        result.map_err(SdkLayerError::DownloadArchive)?;
                        let _ = log_background_bullet.done();
                        break;
                    }
                }
            }

            print::sub_bullet("Verifying SDK checksum");
            verify_checksum(&artifact.checksum, &tarball_path)?;

            print::sub_bullet("Installing SDK");
            decompress_tarball(
                &mut File::open(&tarball_path).map_err(SdkLayerError::OpenArchive)?,
                sdk_layer.path(),
            )
            .map_err(SdkLayerError::DecompressArchive)?;
        }
    }

    Ok(sdk_layer)
}

fn verify_checksum<D>(checksum: &Checksum<D>, path: impl AsRef<Path>) -> Result<(), SdkLayerError>
where
    D: Digest,
{
    let calculated_checksum = fs::read(path.as_ref())
        .map(|data| D::digest(data).to_vec())
        .map_err(SdkLayerError::ReadArchive)?;

    if calculated_checksum == checksum.value {
        Ok(())
    } else {
        Err(SdkLayerError::VerifyArchiveChecksum {
            expected: checksum.value.clone(),
            actual: calculated_checksum,
        })
    }
}

#[derive(Debug)]
pub(crate) enum SdkLayerError {
    DownloadArchive(libherokubuildpack::download::DownloadError),
    DecompressArchive(std::io::Error),
    VerifyArchiveChecksum { expected: Vec<u8>, actual: Vec<u8> },
    OpenArchive(std::io::Error),
    ReadArchive(std::io::Error),
}

impl From<SdkLayerError> for libcnb::Error<DotnetBuildpackError> {
    fn from(value: SdkLayerError) -> Self {
        libcnb::Error::BuildpackError(DotnetBuildpackError::SdkLayer(value))
    }
}
