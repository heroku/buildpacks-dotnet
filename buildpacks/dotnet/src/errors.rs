use crate::dotnet::target_framework_moniker::ParseTargetFrameworkError;
use crate::dotnet::{project, solution};
use crate::dotnet_buildpack_configuration::{
    DotnetBuildpackConfigurationError, ExecutionEnvironmentError,
};
use crate::layers::sdk::SdkLayerError;
use crate::DotnetBuildpackError;
use bullet_stream::{style, Print};
use indoc::formatdoc;
use std::io::{self, stderr};

pub(crate) fn on_error(error: libcnb::Error<DotnetBuildpackError>) {
    match error {
        libcnb::Error::BuildpackError(buildpack_error) => on_buildpack_error(&buildpack_error),
        libcnb_error => log_error(
            "Heroku .NET Buildpack internal buildpack error",
            formatdoc! {"
                The framework used by this buildpack encountered an unexpected error.

                If you can’t deploy to Heroku due to this issue, check the official Heroku Status page at
                status.heroku.com for any ongoing incidents. After all incidents resolve, retry your build.

                Use the debug information above to troubleshoot and retry your build. If you think you found a
                bug in the buildpack, reproduce the issue locally with a minimal example and file an issue here:
                https://github.com/heroku/buildpacks-dotnet/issues/new
            "},
            Some(libcnb_error.to_string()),
        ),
    }
}

#[allow(clippy::too_many_lines)]
fn on_buildpack_error(error: &DotnetBuildpackError) {
    match error {
        DotnetBuildpackError::BuildpackDetection(io_error) => log_io_error(
            "Error completing buildpack detection",
            "determining if we must run the Heroku .NET buildpack for this application.",
            io_error,
        ),
        DotnetBuildpackError::NoSolutionProjects(solution_path) => {
            log_error(
                "No project references found in solution",
                formatdoc! {"
                The solution file `{}` has no project references.

                This buildpack prefers building a solution file over a project file if both
                are present in the root directory.

                To resolve this issue,
                * Delete the solution file to build a root project file instead.
                * Or reference the projects to build from the solution file.

                ", solution_path.to_string_lossy()},
                None,
            );
        }
        DotnetBuildpackError::MultipleRootDirectoryProjectFiles(project_file_paths) => log_error(
            "Multiple .NET project files",
            formatdoc! {"
                The root directory contains multiple .NET project files: `{}`.

                We don’t support having multiple project files in the root directory to prevent
                unexpected results. We recommend reorganizing the directory and project
                structure to include only one project file per folder.

                If you’re porting an application from .NET Framework to .NET, or compiling both
                side-by-side, see Microsoft’s documentation for project organization guidance:
                https://learn.microsoft.com/en-us/dotnet/core/porting/project-structure
                ", project_file_paths.iter()
                    .map(|f| f.to_string_lossy().to_string())
                    .collect::<Vec<String>>()
                    .join("`, `"),
            },
            None,
        ),
        DotnetBuildpackError::LoadSolutionFile(error) => match error {
            solution::LoadError::ReadSolutionFile(io_error) => log_io_error(
                "Error loading solution file",
                "reading the solution file",
                io_error,
            ),
            solution::LoadError::LoadProject(load_project_error) => {
                on_load_dotnet_project_error(load_project_error, "reading solution project files");
            }
        },
        DotnetBuildpackError::LoadProjectFile(error) => {
            on_load_dotnet_project_error(error, "reading root project file");
        }
        DotnetBuildpackError::ParseTargetFrameworkMoniker(error) => match error {
            ParseTargetFrameworkError::InvalidFormat(tfm)
            | ParseTargetFrameworkError::UnsupportedOSTfm(tfm) => {
                log_error(
                    "Unsupported target framework",
                    formatdoc! {"
                        The detected target framework moniker `{tfm}` is either invalid or unsupported. This
                        buildpack currently supports the following TFMs: `net5.0`, `net6.0`, `net7.0`, `net8.0`.

                        For more information, see:
                        https://github.com/heroku/buildpacks-dotnet#net-version
                    "},
                    None,
                );
            }
        },
        DotnetBuildpackError::ReadGlobalJsonFile(error) => log_io_error(
            "Error reading `global.json` file",
            "detecting SDK version requirement",
            error,
        ),
        DotnetBuildpackError::ParseGlobalJson(error) => log_error(
            "Invalid `global.json` file",
            formatdoc! {"
                We can’t parse the root directory `global.json` file because it contains invalid JSON.

                Use the debug information above to troubleshoot and retry your build.
            "},
            Some(error.to_string()),
        ),
        // TODO: Consider adding more specific errors for the parsed values (e.g. an invalid rollForward value)
        DotnetBuildpackError::ParseGlobalJsonVersionRequirement(error) => log_error(
            "Error parsing `global.json` version requirement",
            formatdoc! {"
                We can’t parse the .NET SDK version requirement.

                Use the debug information above to troubleshoot and retry your build. For more
                information, see:
                https://github.com/heroku/buildpacks-dotnet#net-version
            "},
            Some(error.to_string()),
        ),
        DotnetBuildpackError::ParseInventory(error) => log_error(
            "Invalid `inventory.toml` file",
            formatdoc! {"
                We can’t parse the inventory of .NET SDK releases. This error
                is almost always a buildpack bug.

                If you see this error, please file an issue here:
                https://github.com/heroku/buildpacks-dotnet/issues/new

            "},
            Some(error.to_string()),
        ),
        DotnetBuildpackError::ParseSolutionVersionRequirement(error) => log_error(
            "Invalid .NET SDK version requirement",
            formatdoc! {"
                We can’t parse the inferred .NET SDK version requirement.

                Use the debug information above to troubleshoot and retry your build. If you think
                you found a bug in the buildpack, reproduce the issue locally with a minimal
                example and file an issue here:
                https://github.com/heroku/buildpacks-dotnet/issues/new

            "},
            Some(error.to_string()),
        ),
        DotnetBuildpackError::ResolveSdkVersion(version_req) => log_error(
            "Unsupported .NET SDK version",
            formatdoc! {"
                We can’t find a compatible .NET SDK release for the detected version
                requirement ({version_req}).

                For a complete inventory of supported .NET SDK versions and platforms, see:
                https://github.com/heroku/buildpacks-dotnet/blob/main/buildpacks/dotnet/inventory.toml
            "},
            None,
        ),
        DotnetBuildpackError::SdkLayer(error) => match error {
            SdkLayerError::DownloadArchive(error) => log_error(
                "Failed to download .NET SDK",
                formatdoc! {"
                    An unexpected error occurred while downloading the .NET SDK. This error can occur
                    due to an unstable network connection.

                    Use the debug information above to troubleshoot and retry your build.
                "},
                Some(error.to_string()),
            ),
            SdkLayerError::ReadArchive(io_error) => {
                log_io_error(
                    "Error reading downloaded SDK archive",
                    "calculating checksum for the downloaded .NET SDK archive",
                    io_error,
                );
            }
            SdkLayerError::VerifyArchiveChecksum { expected, actual } => log_error(
                "Corrupted .NET SDK download",
                formatdoc! {"
                    Validation of the downloaded .NET SDK failed due to a checksum mismatch. This error can
                    occur intermittently.

                    Use the debug information above to troubleshoot and retry your build. If the issue persists,
                    file an issue here:
                    https://github.com/heroku/buildpacks-dotnet/issues/new

                    Expected: {expected}
                    Actual: {actual}
                ", expected = hex::encode(expected), actual = hex::encode(actual) },
                None,
            ),
            SdkLayerError::OpenArchive(io_error) => {
                log_io_error(
                    "Error reading downloaded SDK archive",
                    "decompressing downloaded .NET SDK archive",
                    io_error,
                );
            }
            SdkLayerError::DecompressArchive(io_error) => log_io_error(
                "Failed to decompress .NET SDK",
                "extracting .NET SDK archive contents",
                io_error,
            ),
        },
        DotnetBuildpackError::ParseBuildpackConfiguration(error) => match error {
            DotnetBuildpackConfigurationError::InvalidMsbuildVerbosityLevel(verbosity_level) => {
                log_error(
                    "Invalid MSBuild verbosity level",
                    formatdoc! {"
                        The `MSBUILD_VERBOSITY_LEVEL` environment variable value (`{verbosity_level}`)
                        is invalid. Did you mean one of the following supported values?

                        d
                        detailed
                        diag
                        diagnostic
                        m
                        minimal
                        n
                        normal
                        q
                        quiet
                    "},
                    None,
                );
            }
            DotnetBuildpackConfigurationError::ExecutionEnvironmentError(error) => match error {
                ExecutionEnvironmentError::UnsupportedExecutionEnvironment(
                    execution_environment,
                ) => {
                    log_error(
                        "Unsupported execution environment'",
                        formatdoc! {"
                            The `CNB_EXEC_ENV` environment variable value (`{execution_environment}`)
                            is not supported. This buildpack currently supports `production` and
                            `test` execution environments.
                        "},
                        None,
                    );
                }
            },
        },
        DotnetBuildpackError::RestoreDotnetToolsCommand(error) => match error {
            fun_run::CmdError::SystemError(_message, io_error) => log_io_error(
                "Unable to restore .NET tools",
                "running the command to restore .NET tools",
                io_error,
            ),
            fun_run::CmdError::NonZeroExitNotStreamed(output)
            | fun_run::CmdError::NonZeroExitAlreadyStreamed(output) => log_error(
                "Unable to restore .NET tools",
                formatdoc! {"
                    The `dotnet tool restore` command exited unsuccessfully ({exit_status}).

                    This error usually happens due to configuration errors. Use the command output
                    above to troubleshoot and retry your build.

                    Restoring .NET tools can also fail for a number of other reasons, such as
                    intermittent network issues, unavailability of the NuGet package feed and/or
                    other external dependencies, etc.

                    Try again to see if the error resolves itself.
                ", exit_status = output.status()},
                None,
            ),
        },
        DotnetBuildpackError::PublishCommand(error) => match error {
            fun_run::CmdError::SystemError(_message, io_error) => log_io_error(
                "Unable to publish",
                "running the command to publish the .NET solution/project",
                io_error,
            ),
            fun_run::CmdError::NonZeroExitNotStreamed(output)
            | fun_run::CmdError::NonZeroExitAlreadyStreamed(output) => log_error(
                "Unable to publish",
                formatdoc! {"
                    The `dotnet publish` command exited unsuccessfully ({exit_status}).

                    This error usually happens due to compilation errors. Use the command output
                    above to troubleshoot and retry your build.

                    The publish process can also fail for a number of other reasons, such as
                    intermittent network issues, unavailability of the NuGet package feed and/or
                    other external dependencies, etc.

                    Try again to see if the error resolves itself.
                ", exit_status = output.status()},
                None,
            ),
        },
        DotnetBuildpackError::CopyRuntimeFiles(io_error) => log_io_error(
            "Error copying .NET runtime files",
            "copying .NET runtime files from the SDK layer to the runtime layer",
            io_error,
        ),
    }
}

fn on_load_dotnet_project_error(error: &project::LoadError, occurred_while: &str) {
    match error {
        project::LoadError::ReadProjectFile(io_error) => {
            log_io_error("Error loading the project file", occurred_while, io_error);
        }
        project::LoadError::XmlParseError(xml_parse_error) => log_error(
            "Error parsing the project file",
            formatdoc! {"
                We can’t parse the project file’s XML content. Parsing errors usually
                indicate an error in the project file.
                
                Use the debug information above to troubleshoot and retry your build."},
            Some(xml_parse_error.to_string()),
        ),
        project::LoadError::MissingTargetFramework(project_path) => {
            log_error(
                "Project file missing TargetFramework property",
                formatdoc! {"
                    The project file `{project_path}` is missing the `TargetFramework` property.
                    You must set this required property.

                    For more information, see:
                    https://github.com/heroku/buildpacks-dotnet#net-version
                ", project_path = project_path.to_string_lossy()},
                None,
            );
        }
    }
}

fn log_io_error(header: &str, occurred_while: &str, io_error: &io::Error) {
    log_error(
        header,
        formatdoc! {"
            An unexpected I/O error occurred while {occurred_while}.

            Use the debug information above to troubleshoot and retry your build. If the
            issue persists, file an issue here:
            https://github.com/heroku/buildpacks-dotnet/issues/new
        "},
        Some(io_error.to_string()),
    );
}

fn log_error(header: impl AsRef<str>, body: impl AsRef<str>, error: Option<String>) {
    let mut log = Print::new(stderr()).without_header();
    if let Some(error) = error {
        let bullet = log.bullet(style::important("Debug info"));
        log = bullet.sub_bullet(error).done();
    }
    log.error(formatdoc! {"
        {header}

        {body}
    ", header = header.as_ref(), body = body.as_ref(),
    });
}
