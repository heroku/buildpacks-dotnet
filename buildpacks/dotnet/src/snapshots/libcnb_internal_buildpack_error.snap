---
source: buildpacks/dotnet/src/errors.rs
---
- [1;36mDebug info[0m
  - Couldn't create platform from platform path: foo bar baz

[0;31m! Heroku .NET Buildpack internal buildpack error[0m
[0;31m![0m
[0;31m! The framework used by this buildpack encountered an unexpected error.[0m
[0;31m![0m
[0;31m! If you canâ€™t deploy to Heroku due to this issue, check the official Heroku Status page at[0m
[0;31m! status.heroku.com for any ongoing incidents. After all incidents resolve, retry your build.[0m
[0;31m![0m
[0;31m! Use the debug information above to troubleshoot and retry your build. If you think you found a[0m
[0;31m! bug in the buildpack, reproduce the issue locally with a minimal example and file an issue here:[0m
[0;31m! https://github.com/heroku/buildpacks-dotnet/issues/new[0m
